### PR TITLE
Fix Reshuffle implementation in Java SDK

### PR DESCRIPTION
**Pull request by kennknowles**
This fixes the Reshuffle default implementation to preserve all metadata and be a true no-op, whereas before it did not preserve the pane info.

Regarding update compatibility, my understanding of runners that have long-lived streaming jobs and perform updates, as of this writing:

 - Flink: overrides Reshuffle
 - Samza: overrides Reshuffle
 - Dataflow v1: overrides Reshuffle
 - Dataflow v2: executes default implementation

I went ahead and used the update compatibility flag anyhow just in case, so it can be reverted to exactly the old implementation on demand.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @robertwb

**Comment by kennknowles** [comment]
Failure was `:runners:spark:3:test` https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/28710/testReport/junit/org.apache.beam.runners.spark.translation.streaming/ResumeFromCheckpointStreamingTest/testWithResume/

**Comment by kennknowles** [comment]
run java precommit

**Comment by kennknowles** [comment]
The failure in the AWS2 tests was the same Spark resume from checkpoint test. It shouldn't even have been run in that suite. It looks like our invocations are a bit messed up (@damccorm FYI but I'm assuming this is just the same as it was in Jenkins).

**Comment by kennknowles** [comment]
run java precommit

**Comment by kennknowles** [comment]
I get the impression this somehow actually slows things down. Everything looks like a timeout. Or perhaps it just invalidates too much of the Gradle cache?

**Comment by kennknowles** [comment]
https://ge.apache.org/s/s6jldnngfhe32/tests/overview?outcome=FAILED#:runners:java-fn-execution:test-5-org.apache.beam.runners.fnexecution.control.remoteexecutiontest-11-testbundleprocessorthrowsexecutionexceptionwhenusercodethrows-44

I haven't figured out if this is flaky or environmental. I can't repro it locally.

**Comment by kennknowles** [comment]
run java precommit

**Comment by kennknowles** [comment]
@robertwb finally green - I don't think this accomplishes everything from the dev thread, just the SDK implementation. It does not impact runners that override or specially translate reshuffle.

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
run flink validatesrunner

**Comment by kennknowles** [comment]
run spark validatesrunner

**Comment by kennknowles** [comment]
run samza validatesrunner

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java [hunk 1]
```diff
@@ -81,28 +88,107 @@ public PCollection<KV<K, V>> expand(PCollection<KV<K, V>> input) {
...
+  private static class RestoreTimestamps<K, V>
+      extends PTransform<PCollection<KV<K, ValueInSingleWindow<V>>>, PCollection<KV<K, V>>> {
+    @Override
+    public PCollection<KV<K, V>> expand(PCollection<KV<K, ValueInSingleWindow<V>>> input) {
+      return input.apply(
+          ParDo.of(
+              new DoFn<KV<K, ValueInSingleWindow<V>>, KV<K, V>>() {
+                @Override
+                public Duration getAllowedTimestampSkew() {
+                  return Duration.millis(Long.MAX_VALUE);
```
Comment (robertwb) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java:181]: The runner doesn't read/act on this value anywhere, does it?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java [hunk 2]
```diff
@@ -81,28 +88,107 @@ public PCollection<KV<K, V>> expand(PCollection<KV<K, V>> input) {
...
+                    for (ValueInSingleWindow<V> value : element.getValue()) {
                       r.output(KV.of(key, value));
                     }
                   }
                 }))
-        .apply("RestoreOriginalTimestamps", ReifyTimestamps.extractFromValues());
+        .apply(
+            "RestoreOriginalWindows",
+            Window.into(new RestoreWindowsFn<>(originalStrategy.getWindowFn())))
+        .apply("RestoreOriginalTimestamps", new RestoreTimestamps<>())
```
Comment (robertwb) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java:116]: What about the PaneInfo information?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java [hunk 3]
```diff
@@ -81,28 +88,107 @@ public PCollection<KV<K, V>> expand(PCollection<KV<K, V>> input) {
...
-        .apply(rewindow)
-        .apply("ReifyOriginalTimestamps", Reify.timestampsInValue())
-        .apply(GroupByKey.create())
-        // Set the windowing strategy directly, so that it doesn't get counted as the user having
-        // set allowed lateness.
-        .setWindowingStrategyInternal(originalStrategy)
+    PCollection<KV<K, ValueInSingleWindow<V>>> reified =
+        input
+            .apply("SetIdentityWindow", rewindow)
+            .apply("ReifyOriginalMetadata", Reify.windowsInValue());
```
Comment (robertwb) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java:94]: Won't this break backwards compatibility for everything using a shuffle (though I concede it's fixing a bug)?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java (hunk 1 above)
Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java:181]: Noting that this is not a change but a move. But no the runner does not act on it. In fact it is unsafe for this reason - the runner may well drop data that you have allowed to be output.

**Comment by kennknowles** [comment]
OK, I added the capability to `outputWindowedValue` to the SDK (somewhat naively IDE-guided refactor) and used it in `Create` in order to write a basic test of what I think we want.

**Comment by kennknowles** [comment]
Gotta go through and poke a few more files

**Comment by robertwb** [comment]
What's the status on this?

**Comment by kennknowles** [comment]
java precommit timed out - perhaps due to cache invalidation due to modifying core SDK

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java (hunk 3 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java:88]: Could we test that this does not break compatibility (even if the test is a one-off manual test)?

**Comment by kennknowles** [comment]
The `outputWindowedValue` change has been merged independently, shrinking this one. I'll check it over to make sure the residual change makes sense and is adequately complete.

**Comment by kennknowles** [comment]
Failure is the error on `master` fixed by #29798 (using `\n` instead of `%n` in a format string)

**Comment by kennknowles** [comment]
Noting that this was green but I went ahead and added the old expansion as an option using the new update compat flag. I am going to use the flag to test compatibility on Dataflow v1. We expect the override makes this change irrelevant, so we don't want to actually bug users. And on v2 we want the incompatibility.

**Comment by kennknowles** [comment]
Added internal Dataflow test that reloads the pipeline with changing the compatibility flag.

**Comment by Abacn** [comment]
It seems the newly added test `testReshufflePreservesMetadata` not working for all runners:

beam_PostCommit_Java_PVR_Samza:  https://github.com/apache/beam/actions/runs/7399747764/job/20131866953

beam_PostCommit_Java_PVR_Spark3_Streaming: https://github.com/apache/beam/actions/runs/7399737943/job/20131836389

beam_PostCommit_Java_PVR_Spark3_Batch: https://github.com/apache/beam/runs/20340841429

beam_PostCommit_Java_ValidatesRunner_ULR also

**Pull request merged**